### ISSUE TITLE
How Yolov5 calculate mAP@.5 and mAP@.5:.95 and plot images?

### ISSUE THREAD

**Issue by Github-Vicente** [ultralytics/yolov5#4052]
## ❔Question
I have three question:

1. When I am using `val.py`, no matter what the `--iou-thres` and `--conf-thres` I set. The `test_batch0_pred.jpg` will plot the bbox only with bbox conf-thres greater then default value 0.25. 
So if I set the `--iou-thres` and `--conf-thres` by myself, it seems that mAP calculation and plot function will under different criteria. Is it normal?
While the record in txt did follow the `--conf-thres` I set. But `test_batch0_pred.jpg`  is not.
Where can I control the threshold of plot iou?

2. Sometimes val.py output a `test_batch0_pred.jpg` with mosaic image. Did anyone face this problem too?

3. Is it normal that I apply `val.py` and `detect.py` without any customize parameter to the same dataset, but get different result?

Thank you!

**Comment by glenn-jocher** [comment]
@GamuzaLu 👋 Hello, thank you for asking about the differences between [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py), [detect.py](https://github.com/ultralytics/yolov5/blob/master/detect.py) and [test.py](https://github.com/ultralytics/yolov5/blob/master/test.py) in YOLOv5.

These 3 files are designed for different purposes and utilize different dataloaders with different settings. train.py dataloaders are designed for a speed-accuracy compromise, test.py is designed to obtain the best mAP on a validation dataset, and detect.py is designed for best real-world inference results. A few important aspects of each:

## `train.py`

- **trainloader** LoadImagesAndLabels(): designed to load train dataset images and labels. Augmentation capable and enabled. https://github.com/ultralytics/yolov5/blob/fca5e2a48fb526b57bda0c66be6b7ac1aaa8d83d/train.py#L188-L192
- **testloader** LoadImagesAndLabels(): designed to load val dataset images and labels. Augmentation capable but disabled. https://github.com/ultralytics/yolov5/blob/fca5e2a48fb526b57bda0c66be6b7ac1aaa8d83d/train.py#L199-L202
- image size: `640`
- rectangular inference: `False`
- [confidence threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L291): `0.001`
- [iou threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L292): `0.6`
- [multi-label](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L120): `True`
- [padding](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L83-L90): `None`

## `test.py`

- **dataloader** LoadImagesAndLabels(): designed to load train, val, test dataset images and labels. Augmentation capable but disabled. https://github.com/ultralytics/yolov5/blob/fca5e2a48fb526b57bda0c66be6b7ac1aaa8d83d/test.py#L89-L90
- image size: `640`
- rectangular inference: `True`
- [confidence threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L291): `0.001`
- [iou threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L292): `0.6`
- [multi-label](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L120): `True`
- [padding](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L83-L90): `0.5 * maximum stride`

## `detect.py`

- **dataloaders** (multiple): designed for loading multiple types of media (images, videos, globs, directories, streams). https://github.com/ultralytics/yolov5/blob/fca5e2a48fb526b57bda0c66be6b7ac1aaa8d83d/detect.py#L46-L53
- image size: `640`
- rectangular inference: `True`
- [confidence threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/detect.py#L152): `0.25` 
- [iou threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/detect.py#L153): `0.45`
- [multi-label](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/utils/general.py#L415): `False`
- [padding](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/detect.py#L45-L54): `None`

## YOLOv5 PyTorch Hub

`models.autoShape()` class used for image loading, preprocessing, inference and NMS. For more info see [YOLOv5 PyTorch Hub Tutorial](https://docs.ultralytics.com/yolov5/tutorials/pytorch_hub_model_loading) https://github.com/ultralytics/yolov5/blob/fca5e2a48fb526b57bda0c66be6b7ac1aaa8d83d/models/common.py#L225-L250

- image size: `640`
- rectangular inference: `True`
- [confidence threshold](https://github.com/ultralytics/yolov5/blob/abfcf9eb79877971acd238cafe6149711c5056ad/models/common.py#L226-L230): `0.25 `
- [iou threshold](https://github.com/ultralytics/yolov5/blob/abfcf9eb79877971acd238cafe6149711c5056ad/models/common.py#L226-L230): `0.45`
- [multi-label](https://github.com/ultralytics/yolov5/blob/abfcf9eb79877971acd238cafe6149711c5056ad/models/common.py#L288): `False`
- [padding](https://github.com/ultralytics/yolov5/blob/abfcf9eb79877971acd238cafe6149711c5056ad/models/common.py#L226-L230): `None`
